Fix mapping bug

MorganRodgers opened the pull request:
Includes fixes for #4 and #5.

Commit:
Remove test for deleted controller

Commit:
Addition of tests for mapping and fixes for FACL logic

Commit:
Pathname related changes

Made APP_ROOT a constant
Switched to Pathname for all path related operations

Commit:
Fixes for issue #4

Change Mapping.app to be a Pathname
Removed Mapping lifecycle hooks since a custom save/destroy method was being used anyway
Re-wrote FACL logic to take advantage of Mapping.app and Mapping.dataset being Pathnames
Re-wrote tests using new FACL logic

Commit:
Fixes for #5

Simplified the implementations of app_list and known_datasets

Commit:
Fixes for #5

Renamed APP_PROJECT_SPACE to SHARED_APPS_ROOT to follow convention

ericfranz commented on app/helpers/mappings_helper.rb:
```diff
@@ -34,11 +27,7 @@ def get_app_name app_path
-        ENV['APP_DATASET_ROOT']
-      ), '*'
-    ).sort
+    installed_datasets = Dir.glob(Configuration.app_dataset_root.join('*')).sort
```
If you make this a separate method then both become functional:

```ruby
def managed_datasets
  Dir.glob(Configuration.app_dataset_root.join('*')).sort
end

def all_datasets
  managed_datasets | Mapping.datasets
end
```

ericfranz commented on app/models/mapping.rb:
```diff
@@ -1,13 +1,22 @@
+  validates :user, :app, :dataset, presence: true
+  YAML_FILE_PATH = Pathname.new(ENV['SHARED_APPS_ROOT']).expand_path().join('mappings.yaml')
+  APP_ROOT = Pathname.new(ENV['SHARED_APPS_ROOT']).expand_path()
+  FACL_USER_DOMAIN = ENV['FACL_USER_DOMAIN']
```
I'd make these methods i.e. `yaml_file_path` and `app_root` and `facl_user_domain`. Yes it is slightly slower but not noticeable, and IMHO the benefits are: improves the readability a little bit, and if it came to it, makes mocking easier and overriding easier. Also these references to ENV can also be replaced with calls to `Configuration` methods.

As soon as you do this you will recognize that Mapping model now has Configuration calls embedded in it. Not my favorite, as it couples the two together and requires mocking the Configuration but at least isolating the Configuraiton method calls to separate methods is better. We just introduced the ConfigurationSingleton/Configuration class/instances before you came to all of our apps - long story will share later. But I just started seeing the user of this in different places as a result, which is undesirable - but the solution to me isn't clear, because we can't just put it in the initializer because we like to use Rails ActiveModel class methods to initialize these instances...

MorganRodgers replied:
I agree that if we are going to use a Configuration singleton then we ought to use it consistently. I prefer using constants for environment variables that should remain unchanged for an entire request. I don't find testing and mocking a compelling argument as we already have the ability to set separate values for testing, development and production instances.

ericfranz commented on app/models/mapping.rb:
```diff
@@ -73,68 +84,66 @@ def save_and_set_facls
+  def add_rx_facl(pathname)
+    unless should_add_facl?(pathname)
+      return
+    end
```
For idomatic ruby, make this block a single line of code:

```ruby
def add_rx_facl(pathname)
  return unless should_add_facl?(pathname)
```

ericfranz commented on app/models/mapping.rb:
```diff
@@ -73,68 +84,66 @@ def save_and_set_facls
-  after_create do |mapping|
-    Mapping.dump_to_yaml
+    return true
   end
```
This could be simplified without losing readability, and comments added to right or on top:

```ruby
def should_add_facl?(pathname)
  pathname.owned? && ! rx_facl_exists?(pathname) # note: calling owned? before rx_facl_exists? protects rx_facl_exists? from throwing InvalidPath
end
```
This is easier to read at a glance and see what is happening.

I think that while its interesting that `# Using writable would try setting a FACL on /dev/null...`  in this case even if it didn't the logic would still be preferred to only set the facl if you own the file (thus testing for ownership). I might have write access to a client's dataset because his permissions were generously set, but its still his dataset so his responsibility to manage access.

I know some people will go as far as to actually use refinements, which is scoped monkey patching, to add methods these methods on Pathname itself i.e. `Pathname#rx_facl_exists?` which then gives you the ability to write:

```ruby
def should_add_facl?(pathname)
  pathname.owned? && ! pathname.rx_facl_exists?
end
```

But the problem is that those refinements are only available in the place where we explicitly choose to use the refinement. And unless I know the Pathname method list on the back of my hand, I might try to use that cool new Pathname method somewhere else and get a "method missing error" because I didn't realize it was defined in a refinement.

So IMHO what you have done, defining `should_add_facl?` and `rx_facl_exists?` as methods that accept pathname is perfect. 

And you made the right call: isolating the logic under test `should_add_facl?` from the method that has side effects `add_rx_facl` is exactly what you should do.

MorganRodgers replied:
I find the idea of classes being open for modification without having to subclass in Ruby odd, and I don't think that it is something that I'll take up as a practice.

I do quite like the ability to use question marks in method names (or exclamation marks).

ericfranz replied:
That classes are open for modification does have some really great use cases, but have obvious caveats. Some benefits:

1. If you stick to a single Ruby extensions library, like ActiveSupport, all of the extra methods added to base Ruby classes make life really easy. You can't mix large extension libraries unless the authors work hard to not introduce conflicts of course.
2. It makes writing mocking libraries and testing frameworks easier.
3. In Open OnDemand, we add paths under /etc/ood/config to the list of paths Rails looks for initializers, which means that at any time any admin could drop an initializer in that monkey patches (and thus changes) any method on any class in an OnDemand app. Which may be a good or really really bad idea; almost any defect can be temporarily addressed using this approach (I need a fix now can't wait till the next OOD release type of thing). Of course that is a step above modifying the code inline of an app to make it work.
4. It is actually how ActiveRecord and Rails plugins work ([osc_machete_rails](https://github.com/OSC/osc_machete_rails/blob/103bb9b37b40684d4745b7198ebab854a5f6121e/lib/osc_machete_rails/workflow.rb#L246-L247)). We also borrowed the Rails approach to constructing the correct Adapter subclass for databases based on a configuration file dynamically to do so for OodCore job adapters: each adapter creates a "build_adapter" method on the Adapter superclass and a single factory will execute this method and pass the rest of the configuration to this method:

https://github.com/OSC/ood_core/blob/d21a1d62a41b1a3e6d05a28cfe1100bb931597f9/lib/ood_core/job/adapters/lsf.rb#L5-L19

https://github.com/OSC/ood_core/blob/d21a1d62a41b1a3e6d05a28cfe1100bb931597f9/lib/ood_core/job/factory.rb#L30-L36

Though IMHO that is overkill - ood_core is not Rails, we don't need that much flexibility. I'll open an issue to simplify that code.

ericfranz commented on app/models/mapping.rb:
```diff
@@ -149,63 +158,44 @@ def build_facl_entry_for_user(user, domain)
-    )
+  # Check whether a user has read/execute permissions on the app and dataset directories
+  # @return [Boolean] does user have correct permissions?
+  def user_has_permissions_on_both
```
Is this still a necessary method now that we have `rx_facl_exists?`? Seems like the body of this method duplicates the purpose of `rx_facl_exists?`.

MorganRodgers replied:
I'd like to leave it in as a shortcut, but calling `Mapping.rx_facl_exists?` instead of duplicating the work. Also since it's a Boolean method I'll add a question mark.

MorganRodgers replied:
Actually I spoke too soon. They do different things. `user_has_permissions_on_both` ensures that a user is `allow?'d` to do something, whereas `rx_facl_exists?` checks for the entry. A user can have permission to do something without a FACL entry existing.

ericfranz replied:
Oh also when referring to instance methods of a class, use `Mapping#rx_facl_exists?` instead of `Mapping.rx_facl_exists?` as the `Mapping.rx_facl_exists?` form by convention refers to the "class method" instead of the "instance method".

MorganRodgers replied:
Noted.

MorganRodgers replied:
And on third thought I am going to re-write `user_has_permissions_on_both?` to use the Pathname tests as they do not throw.

MorganRodgers replied:
...except that the Pathname methods only test for properties when acting as `$USER`, so we do need the flexibility of the FACL testing.

ericfranz commented:
One caveat of using `Pathname` - those times when you actually need a String, you sometimes have to call `Pathname#to_s`. In this case, when writing to the YAML file:

```
---
:mappings:
- :app: !ruby/object:Pathname
    path: "/users/PZS0562/efranz/awesim/dev/bc_shiny_example"
  :user: efranz
  :dataset: !ruby/object:Pathname
    path: "/users/PZS0562/efranz/awesim/dev/bc_shiny_app_template/data/hista_alignment"
  :extensions:
```

And when accepting a pathname as an argument, sometimes you have to coerce the argument to a string then to a Pathname so you can accept both arguments i.e.:

```ruby
# now my method can handle both strings and Pathname objects
def my_function(pathname)
  pathname = Pathname.new(pathname.to_s)
```

Though I can add protection and coerce to a string after loading the YAML. I guess it begs the question, if all our code is in Ruby, and we want to mostly deal in Ruby objects, is there any problem with keeping the `!ruby/object:Pathname` in the mappings.yaml file? I don't know, maybe not.

MorganRodgers commented:
I think that we should prefer the common denominator. This file is meant as communication between applications. For now they are both written in Ruby, but that may not always be the case. Unless a change to String will derail something you are doing I am going to make that change on my end.

Commit:
Fixes for #6

Changing YAML store to use string for paths instead of Ruby-specific interpretration of Pathname
Various readability changes

ericfranz commented:
String sounds good. Also, while the code I have for the Shiny app itself will handle `:app` in the mapping as either an absolute path (/path/to/bc_shiny_foo) or just a directory name (bc_shiny_foo) it might be better to just use the directory name in the mappings.yml (or even the database). Of course the dataset will remain an absolute path. Reasoning:

1. one less absolute path to persist; if/when things change, those absolute paths cause problems
2. while BISR will have an admin account, could they presumably use a different admin account in the future for sharing apps, etc.?
3. the apps themselves have to know where to look for the mappings.yml file, so right now they look in the app directory or the parent directory

The working assumption right now has been that apps will be in a single directory by convention. But I don't know there may be reasons to keep it an absolute path. It probably does not matter.

Commit:
Simplified should_remove_facl and added more return type comments

Commit:
One lined some conditionals to improve readability

Commit:
Fixes for #5, #6 and #7

Moved configuration into config singleton
Improved validation and error messages
Updated tests to forbid multiple user/app mappings
